File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int l=0,r=1;
        int n=height.size();
        int  trap=0;
        while(l<=r && r<n){
            if(height[r]<height[l]){
                trap+=(height[l]-height[r]);
                r++;
            }
            else {
                l=r;
                r++;
            }
        }
        if(l!=n-1){
            int a=n-1,b=n-2;
            trap-=height[l]-height[a];
              while(b>l){
              trap-=height[l]-height[b];

               if(height[b]<height[a]){
                trap+=height[a]-height[b];
                b--;
            }
            else {
                a=b;
                b--;
            }
              }
        }
        return trap;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        // concept of right and left max
        int leftMax=height[0];
        int rightMax=height[height.length-1];
        int water=0;
        int l=0,r=height.length-1;

        while(l<=r){
            if(leftMax<rightMax){
                leftMax=Math.max(leftMax,height[l]);
                water+=leftMax-height[l];   //left boundary - current tower , obviously right is greater so water is stored
                l++;
            }
            else{
                rightMax=Math.max(rightMax,height[r]);
                water+=rightMax-height[r];  
                r--;
            }
        }

        return water;
    }
}